### PR TITLE
Improve CurrencyEditText

### LINKED ISSUE #5906: Improve CurrencyEditText functionality
The `CurrencyEditText` is not offering the best UX right now:
1. There is no way to differentiate between `0` and empty value. (check if this is the cause of https://github.com/woocommerce/woocommerce-android/issues/3496)
2. The cursor position is lost when editing the price.
3. There is no way to make text changes to a position different than the last one.
4. The `isInitialized` property is broken, making the `liveData` always initialized with `0` regardless of the wanted initial value.

### PR DESCRIPTION
**Pull request by hichamboushaba**
Closes: #5906
Depends on Fluxc PR https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2296

### Description
This PR tries to improve our `CurrencyEditText` and its layout `WCMaterialOutlinedCurrencyEditTextView`, the list of improvements added is:
1. Ability to move the cursor and edit content anywhere, previously the changes always occurred at the last position.
2. Ability to differentiate between `0` and empty state, this behavior is optional through the property `supportsEmptyState`.
3. Support of negative values, this behavior is optional via the property `supportsNegativeValues`
4. The currency handling is done now using the `prefix`/`suffix` properties of `TextInputLayout`, which makes the cursor position more accurate, instead of pointing to the currency when currency is on the right.
5. Support of two edit modes: one with full formating, and one without it. This is useful for making the editing feels more natural for some screens.

This PR doesn't change any behavior to the screens that were using the `CurrencyEditText`, meaning both `supportsEmptyState` and `supportsNegativeValues` are set to false, and on future PRs we can update each screen with the appropriate combination.
I also was planning on updating `ProductPricingFragment` to go back to using `WCMaterialOutlinedCurrencyEditTextView`, since it supports negative values now, but I decided to do it in a separate PR since this was is already big, I'll try to work on this on a separate PR, and try to fix this [bug](https://github.com/woocommerce/woocommerce-android/issues/3496), since now we can differentiate between empty states and `0`.

This is how the field behaves now:

**Simple Payments**

https://user-images.githubusercontent.com/1657201/156342883-2432eb36-32db-4b92-8452-89124cfb606a.mov 

**Other fields (without full-formating)**

https://user-images.githubusercontent.com/1657201/156342848-04dff67c-2905-49d6-bad9-9799d0417d52.mov

And this is how the UI compares in before vs after:

| Before | After |
| ----- | ----- |
| Screen Shot 2022-02-24 at 16 59 47 | Screen Shot 2022-02-24 at 16 49 43 |
| Screen Shot 2022-02-24 at 17 00 06 | Screen Shot 2022-02-24 at 16 49 11 |
| Screen Shot 2022-02-24 at 16 59 13 | Screen Shot 2022-02-24 at 16 50 52 |
| Screen Shot 2022-02-24 at 16 58 59 | Screen Shot 2022-02-24 at 16 50 29|

### Testing instructions
The `WCMaterialOutlinedCurrencyEditTextView` is currently used in "Simple Payments" and Order Creation (shipping and fees), and it's used in refund by amount (not available in the app now), so please check for any regressions in those areas.
And please if possible, test the changes with these different options:
1. Different keyboards, I tested with Gboard and Samsung Keyboard.
2. Different formatting options from WC Core, especially different options of currency positioning (please note that to force refetch the settings, you need to fully restart the app).

A remark: for order creation, please make sure to change the code manually to update `ORDER_CREATION_M2` to `true` to be able to add `shipping` and `fees`, and please note that the `fees` are not saved when added yet.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt [hunk 1]
```diff
@@ -108,4 +164,132 @@ class WCMaterialOutlinedCurrencyEditTextView @JvmOverloads constructor(
     override fun dispatchRestoreInstanceState(container: SparseArray<Parcelable>) {
         super.dispatchThawSelfOnly(container)
     }
+
+    private enum class EditTextLayoutMode {
+        FILL, WRAP
+    }
+}
+
+private class CurrencyEditText @JvmOverloads constructor(
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt]: I made this as private class to avoid any accidental of usage outside of `WCMaterialOutlinedCurrencyEditTextView`

File: WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt [hunk 2]
```diff
@@ -3,87 +3,143 @@ package com.woocommerce.android.widgets
...
         private const val KEY_SUPER_STATE = "WC-OUTLINED-CURRENCY-VIEW-SUPER-STATE"
     }
+
+    private val currencyEditText: CurrencyEditText = CurrencyEditText(context)
+
+    val editText: TextInputEditText
+        get() = currencyEditText
+
+    @Inject lateinit var wcStore: WooCommerceStore
+    @Inject lateinit var selectedSite: SelectedSite
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt]: Please tell me what you think about this approach?
Previously we were depending on calling `initView` with the correct `currency` and `currencyFormatter`, with DI this is done for us automatically, and makes the component more portable and more encapsulated, since the price formatting is a UI logic anyway.
With this change I was able to remove some code that we didn't need anymore from the ViewModels.

File: WooCommerce/src/main/res/layout/dialog_simple_payments.xml [hunk 3]
```diff
@@ -52,17 +52,20 @@
...
         app:layout_constraintStart_toStartOf="parent" />
 
-    <com.woocommerce.android.widgets.CurrencyEditText
+    <com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView
         android:id="@+id/editPrice"
         android:layout_width="match_parent"
         android:layout_height="wrap_content"
+        app:boxBackgroundMode="none"
         android:gravity="center"
-        android:imeOptions="flagNoExtractUi"
```
Comment (hichamboushaba) [inline WooCommerce/src/main/res/layout/dialog_simple_payments.xml:60]: This value is being set automatically by WCMaterialOutlinedCurrencyEditTextView

File: WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt [hunk 4]
```diff
@@ -108,4 +164,132 @@ class WCMaterialOutlinedCurrencyEditTextView @JvmOverloads constructor(
     override fun dispatchRestoreInstanceState(container: SparseArray<Parcelable>) {
         super.dispatchThawSelfOnly(container)
     }
+
+    private enum class EditTextLayoutMode {
+        FILL, WRAP
+    }
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt:178]: This enum is just for differentiating between two modes:
1. The default mode where the editText will take all the available space of the inputLayout (FILL)
2. The second mode that I added, and it's where the editText will use `wrap_content` inside the inputLayout, this is useful for Simple Payments UI. (WRAP)

**Comment by hichamboushaba** [comment]
@nbradbury @shiki @anitaa1990 I requested a review from the three of you due to your involvement with the CurrencyEditText recently 🙏.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt [hunk 5]
```diff
@@ -3,87 +3,146 @@ package com.woocommerce.android.widgets
...
-        currency: String,
-        decimals: Int,
-        currencyFormatter: CurrencyFormatter
-    ) {
-        binding.currencyEditText.initView(currency, decimals, currencyFormatter)
+        siteSettings?.let {
+            val currencySymbol = wcStore.getSiteCurrency(selectedSite.get(), siteSettings.currencyCode)
+            when (siteSettings.currencyPosition) {
+                LEFT, LEFT_SPACE -> prefixText = currencySymbol
+                RIGHT, RIGHT_SPACE -> suffixText = currencySymbol
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt]: We don't take the `SPACE` property into consideration here, I think it's OK since the currency is a separate view from the content now.

File: WooCommerce/src/test/kotlin/com/woocommerce/android/widgets/CurrencyEditTextTest.kt [hunk 6]
```diff
@@ -1,144 +0,0 @@
-package com.woocommerce.android.widgets
-
-import org.junit.Test
-import kotlin.test.assertEquals
-
-class CurrencyEditTextTest {
```
Comment (hichamboushaba) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/widgets/CurrencyEditTextTest.kt:6]: @shiki I removed those tests since the logic has been simplified a lot after extracting the currency symbol to a different view, but please let me know if you think we should keep some of them 🙏

**Comment by shiki** [comment]
@hichamboushaba I noticed that I get this crash when opening Simple Payments:

```
2022-02-24 16:19:56.151 7071-7071/com.woocommerce.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.woocommerce.android.prealpha, PID: 7071
    android.view.InflateException: Binary XML file line #69 in com.woocommerce.android.prealpha:layout/dialog_simple_payments: Binary XML file line #69 in com.woocommerce.android.prealpha:layout/dialog_simple_payments: Error inflating class com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView
    Caused by: android.view.InflateException: Binary XML file line #69 in com.woocommerce.android.prealpha:layout/dialog_simple_payments: Error inflating class com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView
    Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.view.LayoutInflater.createView(LayoutInflater.java:852)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:1004)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:959)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:1121)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1082)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:680)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:532)
        at androidx.fragment.app.Fragment.onCreateView(Fragment.java:1968)
        at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2995)
        at androidx.fragment.app.DialogFragment.performCreateView(DialogFragment.java:492)
        at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:523)
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:261)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:1840)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1764)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1701)
        at androidx.fragment.app.FragmentManager$4.run(FragmentManager.java:488)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.ClassCastException: android.content.res.TypedArray cannot be cast to java.lang.AutoCloseable
        at com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView.<init>(WCMaterialOutlinedCurrencyEditTextView.kt:66)
        at com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView.<init>(WCMaterialOutlinedCurrencyEditTextView.kt:35)
        at com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView.<init>(Unknown Source:11)
        at java.lang.reflect.Constructor.newInstance0(Native Method) 
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343) 
        at android.view.LayoutInflater.createView(LayoutInflater.java:852) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:1004) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:959) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:1121) 
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1082) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:680) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:532) 
        at androidx.fragment.app.Fragment.onCreateView(Fragment.java:1968) 
        at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2995) 
        at androidx.fragment.app.DialogFragment.performCreateView(DialogFragment.java:492) 
        at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:523) 
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:261) 
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:1840) 
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1764) 
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1701) 
        at androidx.fragment.app.FragmentManager$4.run(FragmentManager.java:488) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7660) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
```

Is this a build problem?

File: WooCommerce/src/test/kotlin/com/woocommerce/android/widgets/CurrencyEditTextTest.kt (hunk 6 above)
Comment (shiki) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/widgets/CurrencyEditTextTest.kt:6]: I added these in the context of “we need to fix this so we can release the beta for Simple Payments ASAP” (#5412). You can imagine that I hacked and skipped a lot of things to do just that. 😄 There were many things that I did not understand about `CurrencyEditText` and I didn't have the luxury of time to refactor and question the past decisions. Hence, I added unit tests just to fix one small problem and have some confidence that I didn't make it worse. 😄 

I agree (and my reviewer, @wzieba, does too) that this class should be refactored at some point. So, thank you for doing just that! 🙇 

TLDR No worries about removing these tests. 😆

**Comment by hichamboushaba** [comment]
> @hichamboushaba I noticed that I get this crash when opening Simple Payments:

Thanks for finding this @shiki, this should be fixed now by b024af6.
The cause of the crash is that the type `TypedArray` was [updated](https://developer.android.com/reference/android/content/res/TypedArray#close()) to implement `AutoCloseable` in API 31, so the IDE imported the standard `use` extension that supported `AutoCloseable` interface, while on our case we need the `use` extension that `androidx` adds, I mean `androidx.core.content.res.use`, I fixed the import on the last commit.

**Comment by hichamboushaba** [comment]
I'm changing this back to draft, I'm still not super satisfied with the experience, and I want to spend a bit more time checking if there is anything left to improve before moving it back to review.
But you are free to test the branch and give me any early feedback 🙏

**Comment by shiki** [comment]
> The cause of the crash is that the type TypedArray was [updated](https://developer.android.com/reference/android/content/res/TypedArray#close()) to implement AutoCloseable in API 31, so the IDE imported the standard use extension that supported AutoCloseable interface, while on our case we need the use extension that androidx adds, I mean androidx.core.content.res.use, I fixed the import on the last commit.

I can confirm that the issue is fixed!

Thanks for explaining the issue. It's amazing that the compilation succeeded when in the end it's just gonna crash. 🤷 

I'll test this again when you think it's ready. 🙂

**Comment by hichamboushaba** [comment]
I think I'm just overthinking this, I tried a totally different approach using `InputFilter` which wouldn't have the formatting as we have now, but would allow a more free and natural editing experience, but I found out that our beloved Android platform has still a lot of issues, and one of them that it need hackish ways to support proper decimal separators (check this [question](https://stackoverflow.com/questions/3821539/decimal-separator-comma-with-numberdecimal-inputtype-in-edittext/45384821#45384821)), which is doable, but made the whole solution as not worth it, as I was aiming for a simpler solution (you can check the code I was working on [here](https://github.com/woocommerce/woocommerce-android/commit/38529b8c1bb34caf41d895aeaa96bef3ce7419e8))

So let's mark this as ready for review as it's, and we can discuss it as a team, and see if it offers a good UX or not.

Just FYI, the commit b106eb60bcdcaf7048b6d9ba5e761c6650efc07d is aiming at restoring the original regex that we had before this [PR](https://github.com/woocommerce/woocommerce-android/pull/3589), as it's suspected that this might be the cause of the [issue](https://github.com/woocommerce/woocommerce-android/issues/3651), and generally Android uses the standard digits when editing numbers, so it doesn't feel natural to have "Arabic-hindu numerals" there.

**Comment by hichamboushaba** [comment]
I think I can't help overthinking this 😅, I installed a build using the alternative solution I started and it started growing on me, the editing feels a bit more natural since I can decide whether I want the decimal separator or not.
But at the same time, I feel like the full formatting offered by the current solution is nice too.

If it's not much, I would love your opinion too before going forward with one of the solutions, I uploaded two videos to show how the other solution works in a [draft PR](https://github.com/woocommerce/woocommerce-android/pull/5940), it's not ready for review though.

**Comment by nbradbury** [comment]
> I feel like the full formatting offered by the current solution is nice too

I also like the formatting - this is what I would opt for.

**Comment by shiki** [comment]
> If it's not much, I would love your opinion too before going forward with one of the solutions, I uploaded two videos to show how the other solution works in a https://github.com/woocommerce/woocommerce-android/pull/5940, it's not ready for review though.

I have mixed feelings. I like the full formatting in the context of the Simple Payments UI. But for other ones, it feels weird to me. I like the simplicity of https://github.com/woocommerce/woocommerce-android/pull/5940 for non-SImple Payments UIs. 

I suspect that there is a history of why we do it this way. Maybe this can be discussed more widely or perhaps involve the design team? 🤷 

Side note: I tested this PR and looks great to me. The only nitpick that I have is when I place the cursor after the decimal separator, I cannot delete it or the number before it. The backspace doesn't do anything. 😅

File: WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt (hunk 2 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt]: Nice!

**Comment by hichamboushaba** [comment]
Thanks everyone for your opinion.

> I have mixed feelings. I like the full formatting in the context of the Simple Payments UI. But for other ones, it feels weird to me. I like the simplicity of https://github.com/woocommerce/woocommerce-android/pull/5940 for non-SImple Payments UIs.

Then why not both 😄, I will create a final PR that allows the consumers to decide whether they want the full-formatting or not.

> The only nitpick that I have is when I place the cursor after the decimal separator, I cannot delete it or the number before it. The backspace doesn't do anything. 

I know about this, it's a downside of having the regex to remove the other characters.
This would be tricky to fix it, and I don't think it's a major thing, why the user would want to put the cursor before the decimal separator that he didn't add 🤔, let's keep things as they are, and I can revisit this in the future.

**Comment by shiki** [comment]
> let's keep things as they are, and I can revisit this in the future.

I agree. 🙂

**Comment by shiki** [comment]
> Then why not both 😄, I will create a final PR that allows the consumers to decide whether they want the full-formatting or not.

I forgot to comment about this. 

1. My first response: 🤣 
2. Second response: Sounds good to me!

**Comment by hichamboushaba** [comment]
@shiki @0nko @nbradbury this is now ready for review, I know it's a big PR, sorry for this, the bulk of the logic is in the file `WCMaterialOutlinedCurrencyEditTextView`.

Just some remarks regarding the alternative way for editing, I tried to account for as many cases as possible:
1. It supports the decimal separator configured on the backend and would convert `.` to it if it was entered.
2. When the field doesn't support empty state, entering `0-` will be converted automatically to `-0`, this makes the experience a bit easier, since we can't remove the `0` digit in this case.
3. The leading `0` will be deleted automatically, this is useful also for cases when a non-empty value is not supported, to avoid having something like `012...` when the user starts typing.

Please check the video I added to the description, it shows some of those scenarios.

**Review by 0nko — APPROVED** [review]
I played around with the app and it looks great, @hichamboushaba! 🏅

**Review by nbradbury — APPROVED** [review]
Looks great to me as well! :shipit:

**Review by shiki — APPROVED** [review]
I'm all good with this too. Thank you!
